### PR TITLE
Modernise the project

### PR DESCRIPTION
**Pull request by ChrisLovering**
This updates the following things in the project:

1. We now use Python 3.11
2. We now use Poetry dependency groups
3. black & isort have been added
4. We no longer install Poetry into the built image
    - a new pre-commit hook have been added to ensure the reqs.txt is kept up to date with poetry's deps 
5. All CI actions have been updated
6. CI now uses reusable workflows

On top of this I have formatted the both repo with isort & black and removed the namespace shadowing we have with the alembic path.

**Note** I could not update pytest-asyncio as we currently rely on legacy behaviour, this has been removed in the next verison up. We need to change our fixtures in order to support the newer versions.

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -3,8 +3,8 @@ max-line-length=120
 docstring-convention=all
 import-order-style=pycharm
 application_import_names=api
-exclude=.cache,.venv,.git,alembic/*
-ignore=
+exclude=.cache,.venv,.git,migrations/*
+extend-ignore=
     B311,W503,E226,S311,T000,B008,E712
```
Comment (wookie184) [inline tox.ini:8]:
```suggestion
    B311,W503,E226,S311,T000,B008,E712,
```

File: tox.ini [hunk 2]
```diff
@@ -3,8 +3,8 @@ max-line-length=120
 docstring-convention=all
 import-order-style=pycharm
 application_import_names=api
```
Comment (wookie184) [inline tox.ini:5]: I think these are `flake8-import-order` settings so can be removed
```suggestion
```

Comment (ChrisLovering) [reply tox.ini]: https://github.com/python-discord/code-jam-management/pull/24/commits/a8caf2fe5036465f80964cbb65cb26fcbf60e64b

File: tox.ini (hunk 1 above)
Comment (ChrisLovering) [reply tox.ini]: https://github.com/python-discord/code-jam-management/pull/24/commits/a8caf2fe5036465f80964cbb65cb26fcbf60e64b

**Pull request merged**